internal/services/authorization/role_assignment_resource_base.go

```diff
@@ -0,0 +1,383 @@
+			ValidateFunc:  validation.StringIsNotEmpty,
+		},
+
+		"role_definition_name": {
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: []string{"role_definition_id"},
+			ValidateFunc:  validation.StringIsNotEmpty,
+		},
+
+		"skip_service_principal_aad_check": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
+			Default:  false,
+		},
+
+		"delegated_managed_identity_resource_id": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: azure.ValidateResourceID,
+		},
+
+		"description": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"condition": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			RequiredWith: []string{"condition_version"},
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"condition_version": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			RequiredWith: []string{"condition"},
+			ValidateFunc: validation.StringInSlice([]string{
+				"1.0",
+				"2.0",
+			}, false),
+		},
+	}
+}
+
+func (br roleAssignmentBaseResource) attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"principal_type": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
+	}
+}
+
+func (br roleAssignmentBaseResource) createFunc(resourceName, scope string) sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			roleAssignmentsClient := metadata.Client.Authorization.ScopedRoleAssignmentsClient
+			roleDefinitionsClient := metadata.Client.Authorization.ScopedRoleDefinitionsClient
+			subscriptionClient := metadata.Client.Subscription.Client
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			name := metadata.ResourceData.Get("name").(string)
+
+			var roleDefinitionId string
+			if v, ok := metadata.ResourceData.GetOk("role_definition_id"); ok {
+				roleDefinitionId = v.(string)
+			} else if v, ok := metadata.ResourceData.GetOk("role_definition_name"); ok {
+				roleName := v.(string)
+				roleDefinitions, err := roleDefinitionsClient.List(ctx, commonids.NewScopeID(scope), roledefinitions.ListOperationOptions{Filter: pointer.To(fmt.Sprintf("roleName eq '%s'", roleName))})
+				if err != nil {
+					return fmt.Errorf("loading Role Definition List: %+v", err)
+				}
+
+				if roleDefinitions.Model != nil && len(*roleDefinitions.Model) == 1 && (*roleDefinitions.Model)[0].Id != nil {
+					roleDefinitionId = *(*roleDefinitions.Model)[0].Id
+				} else {
+					return fmt.Errorf("loading Role Definition List: failed to find role '%s'", roleName)
+				}
+			} else {
+				return fmt.Errorf("either 'role_definition_id' or 'role_definition_name' needs to be set")
+			}
```

Comment:
This is personal preference but the multiple layers of if/else make this difficult to understand at first glance - could we refactor this to 
```suggestion
			if v, ok := metadata.ResourceData.GetOk("role_definition_id"); ok {
				roleDefinitionId = v.(string)
			}
			
			if v, ok := metadata.ResourceData.GetOk("role_definition_name"); ok {
				roleName := v.(string)
				roleDefinitions, err := roleDefinitionsClient.List(ctx, commonids.NewScopeID(scope), roledefinitions.ListOperationOptions{Filter: pointer.To(fmt.Sprintf("roleName eq '%s'", roleName))})
				if err != nil {
					return fmt.Errorf("loading Role Definition List for %s: %+v", roleName, err)
				}

				if roleDefinitions.Model == nil || len(*roleDefinitions.Model) != 1 || (*roleDefinitions.Model)[0].Id == nil {
					return fmt.Errorf("loading Role Definition List: failed to find role '%s'", roleName)
				}
				roleDefinitionId = *(*roleDefinitions.Model)[0].Id
			}
			
			if roleDefinitionId == "" {
				return fmt.Errorf("either 'role_definition_id' or 'role_definition_name' needs to be set")
			}
```

Reply:
code updated